Rollup can generate chunks that are non-importable in a module context

### Rollup Version

v2.56.3

### Operating System (or Browser)

MacOS

### Node Version (if applicable)

v14.17.3

### Link To Reproduction

https://replit.com/@DanielRoe1/rollup-repro#src/index.mjs

### Expected Behaviour

I would expect this generated virtual module chunk to be imported successfully. Perhaps the `#` character could be stripped.

### Actual Behaviour

The hash character splits the import (note: only from within an `.mjs` file). It fails with:
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/home/runner/rollup-repro/dist/_virtual_' imported from /home/runner/rollup-repro/dist/index.mjs
```

This is despite the fact that the import (and filename) is `./_virtual_#test.mjs`.

**Note**: I've filed this against `rollup/rollup` as although the particular example is using the virtual module I can imagine a situation where a valid import is structured (say within TS) that outputs to `.mjs` and therefore produces non-runnable JS.

--- a/src/utils/sanitizeFileName.ts
+++ b/src/utils/sanitizeFileName.ts
@@ -1,11 +1,12 @@
 // https://datatracker.ietf.org/doc/html/rfc2396
-const INVALID_CHAR_RE = /[?*:\x00-\x1f\x7f<>#"{}|\^[\]`]/g
+// eslint-disable-next-line no-control-regex
+const INVALID_CHAR_RE = /[?*:\x00-\x1f\x7f<>#"{}|\\^[\]`]/g;
 
-export function sanitizeFileName (name: string): string {
-	const match = /^[a-z]:/i.exec(name)
-	const driveLetter = match ? match[0] : ''
+export function sanitizeFileName(name: string): string {
+	const match = /^[a-z]:/i.exec(name);
+	const driveLetter = match ? match[0] : '';
 
 	// A `:` is only allowed as part of a windows drive letter (ex: C:\foo)
 	// Otherwise, avoid them because they can refer to NTFS alternate data streams.
-	return driveLetter + name.substr(driveLetter.length).replace(INVALID_CHAR_RE, '_')
+	return driveLetter + name.substr(driveLetter.length).replace(INVALID_CHAR_RE, '_');
 }

--- a/src/utils/sanitizeFileName.ts
+++ b/src/utils/sanitizeFileName.ts
@@ -1,6 +1,6 @@
 // https://datatracker.ietf.org/doc/html/rfc2396
 // eslint-disable-next-line no-control-regex
-const INVALID_CHAR_RE = /[?*:\x00-\x1f\x7f<>#"{}|\\^[\]`]/g;
+const INVALID_CHAR_RE = /[?*:\x00-\x1f\x7f<>#"{}|^[\]`]/g;
 
 export function sanitizeFileName(name: string): string {
 	const match = /^[a-z]:/i.exec(name);

--- a/src/utils/sanitizeFileName.ts
+++ b/src/utils/sanitizeFileName.ts
@@ -1,8 +1,11 @@
-export function sanitizeFileName(name: string): string {
-	const match = /^[a-z]:/i.exec(name);
-	const driveLetter = match ? match[0] : '';
+// https://datatracker.ietf.org/doc/html/rfc2396
+const INVALID_CHAR_RE = /[?*:\x00-\x1f\x7f<>#"{}|\^[\]`]/g
+
+export function sanitizeFileName (name: string): string {
+	const match = /^[a-z]:/i.exec(name)
+	const driveLetter = match ? match[0] : ''
 
 	// A `:` is only allowed as part of a windows drive letter (ex: C:\foo)
 	// Otherwise, avoid them because they can refer to NTFS alternate data streams.
-	return driveLetter + name.substr(driveLetter.length).replace(/[\0?*:]/g, '_');
+	return driveLetter + name.substr(driveLetter.length).replace(INVALID_CHAR_RE, '_')
 }

--- a/src/utils/sanitizeFileName.ts
+++ b/src/utils/sanitizeFileName.ts
@@ -1,6 +1,6 @@
 // https://datatracker.ietf.org/doc/html/rfc2396
 // eslint-disable-next-line no-control-regex
-const INVALID_CHAR_RE = /[?*:\x00-\x1f\x7f<>#"{}|^[\]`]/g;
+const INVALID_CHAR_RE = /[\x00-\x1F\x7F<>*#"{}|^[\]`;?:&=+$,]/g;
 
 export function sanitizeFileName(name: string): string {
 	const match = /^[a-z]:/i.exec(name);

--- a/test/chunking-form/samples/sanitize-chunk-names/_config.js
+++ b/test/chunking-form/samples/sanitize-chunk-names/_config.js
@@ -5,7 +5,7 @@ module.exports = {
 		plugins: [
 			{
 				options(options) {
-					options.input = ['\0virtual:entry-1', '\0virtual:entry-2'];
+					options.input = ['\0virtual:entry-1', '\0virtual:entry-2', 'another-[slug]-#result'];
 					return options;
 				},
 				resolveId(id) {

--- a/test/chunking-form/samples/sanitize-chunk-names/_expected/amd/another-_slug_-_result.js
+++ b/test/chunking-form/samples/sanitize-chunk-names/_expected/amd/another-_slug_-_result.js
@@ -1,0 +1,7 @@
+define((function () { 'use strict';
+
+	var another__slug___result = "another-[slug]-#result";
+
+	return another__slug___result;
+
+}));

--- a/test/chunking-form/samples/sanitize-chunk-names/_expected/cjs/another-_slug_-_result.js
+++ b/test/chunking-form/samples/sanitize-chunk-names/_expected/cjs/another-_slug_-_result.js
@@ -1,0 +1,5 @@
+'use strict';
+
+var another__slug___result = "another-[slug]-#result";
+
+module.exports = another__slug___result;

--- a/test/chunking-form/samples/sanitize-chunk-names/_expected/es/another-_slug_-_result.js
+++ b/test/chunking-form/samples/sanitize-chunk-names/_expected/es/another-_slug_-_result.js
@@ -1,0 +1,3 @@
+var another__slug___result = "another-[slug]-#result";
+
+export { another__slug___result as default };

--- a/test/chunking-form/samples/sanitize-chunk-names/_expected/system/another-_slug_-_result.js
+++ b/test/chunking-form/samples/sanitize-chunk-names/_expected/system/another-_slug_-_result.js
@@ -1,0 +1,10 @@
+System.register([], (function (exports) {
+	'use strict';
+	return {
+		execute: (function () {
+
+			var another__slug___result = exports('default', "another-[slug]-#result");
+
+		})
+	};
+}));

